File: backend/src/calorie_track_ai_bot/api/v1/meals.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, Query, Request, status

from ...schemas import (
    MealCreateFromEstimateRequest,
    MealCreateManualRequest,
    MealCreateResponse,
    MealsCalendarResponse,
    MealsListResponse,
    MealUpdate,
    MealWithPhotos,
)
from ...services.db import (
    db_create_meal_from_estimate,
    db_create_meal_from_manual,
    db_delete_meal,
    db_get_meal_with_photos,
    db_get_meals_calendar_summary,
    db_get_meals_with_photos,
    db_update_meal_with_macros,
)
from ...utils.error_handling import handle_api_errors
from .deps import get_authenticated_user_id

router = APIRouter()


@router.post("/meals", response_model=MealCreateResponse)
@handle_api_errors("meal creation")
async def create_meal(
    request: Request, payload: MealCreateManualRequest | MealCreateFromEstimateRequest
):
    if isinstance(payload, MealCreateManualRequest):
        return await db_create_meal_from_manual(payload)

    # For estimate-based meals, we need to get the user_id from headers
    user_id = await get_authenticated_user_id(request)

    return await db_create_meal_from_estimate(payload, user_id)


@router.get("/meals", response_model=MealsListResponse)
@handle_api_errors("meal retrieval")
async def get_meals(
    request: Request,
    date: str | None = Query(None, description="Specific date (YYYY-MM-DD)"),
    start_date: str | None = Query(None, description="Start date for range (YYYY-MM-DD)"),
    end_date: str | None = Query(None, description="End date for range (YYYY-MM-DD)"),
    limit: int = Query(50, ge=1, le=100, description="Maximum number of meals to return"),
) -> MealsListResponse:
    """Get meals with photos and macronutrients.

    Feature: 003-update-logic-for (Multi-photo support)
    Supports:
    - Single date query: ?date=2025-09-30
    - Date range query: ?start_date=2025-09-23&end_date=2025-09-30
    - Default: today's meals
    """
    user_id = await get_authenticated_user_id(request)

    # Parse date parameters
    query_date = None
    query_start = None
    query_end = None

    try:
        if date:
            query_date = datetime.fromisoformat(date).date()
        elif start_date and end_date:
            query_start = datetime.fromisoformat(start_date).date()
            query_end = datetime.fromisoformat(end_date).date()
        else:
            # Default to today
            query_date = datetime.now().date()
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=f"Invalid date format: {e}"
        ) from None

    # Get meals with photos
    meals = await db_get_meals_with_photos(
        user_id=UUID(user_id),
        query_date=query_date,
        start_date=query_start,
        end_date=query_end,
        limit=limit,
    )

    return MealsListResponse(meals=meals, total=len(meals))


@router.get("/meals/calendar", response_model=MealsCalendarResponse)
@handle_api_errors("meals calendar")
async def get_meals_calendar(
    request: Request,
    start_date: str = Query(..., description="Start date (YYYY-MM-DD)"),
    end_date: str = Query(..., description="End date (YYYY-MM-DD)"),
) -> MealsCalendarResponse:
    """Get daily meal summaries for calendar view.

    Feature: 003-update-logic-for (Calendar navigation)
    Returns aggregated nutrition data by date.
    """
    user_id = await get_authenticated_user_id(request)

    # Parse dates
    try:
        query_start = datetime.fromisoformat(start_date).date()
        query_end = datetime.fromisoformat(end_date).date()
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Invalid date format: {e!s}") from e

    # Validate date range (max 1 year)
    if (query_end - query_start).days > 365:
        raise HTTPException(status_code=400, detail="Date range cannot exceed 1 year")

    # Get calendar summary
    calendar_data = await db_get_meals_calendar_summary(
        user_id=UUID(user_id), start_date=query_start, end_date=query_end
    )

    return MealsCalendarResponse(dates=calendar_data)


@router.get("/meals/{meal_id}", response_model=MealWithPhotos)
@handle_api_errors("meal retrieval by ID")
async def get_meal(meal_id: str, request: Request) -> MealWithPhotos:
    """Get a specific meal with photos and macronutrients.

    Feature: 003-update-logic-for (Multi-photo support)
    """
    user_id = await get_authenticated_user_id(request)

    try:
        meal = await db_get_meal_with_photos(UUID(meal_id))
    except ValueError:
        raise HTTPException(status_code=422, detail="Invalid meal ID format") from None

    if not meal:
        raise HTTPException(status_code=404, detail="Meal not found")

    # Verify ownership
    if str(meal.userId) != user_id:
        raise HTTPException(status_code=403, detail="You do not own this meal")

    return meal


@router.patch("/meals/{meal_id}", response_model=MealWithPhotos)
@handle_api_errors("meal update")
async def update_meal(meal_id: str, payload: MealUpdate, request: Request) -> MealWithPhotos:
    """Update meal description or macronutrients.

    Feature: 003-update-logic-for (Meal editing)
    Automatically recalculates calories from macros if updated.
    """
    user_id = await get_authenticated_user_id(request)

    # Validate meal exists and ownership
    try:
        existing_meal = await db_get_meal_with_photos(UUID(meal_id))
    except ValueError:
        raise HTTPException(status_code=422, detail="Invalid meal ID format") from None

    if not existing_meal:
        raise HTTPException(status_code=404, detail="Meal not found")

    if str(existing_meal.userId) != user_id:
        raise HTTPException(status_code=403, detail="You do not own this meal")

    # Update meal with macronutrient recalculation
    updated_meal = await db_update_meal_with_macros(meal_id=UUID(meal_id), updates=payload)

    if not updated_meal:
        raise HTTPException(status_code=500, detail="Failed to update meal")

    return updated_meal


@router.delete("/meals/{meal_id}", status_code=status.HTTP_204_NO_CONTENT)
@handle_api_errors("meal deletion")
async def delete_meal(meal_id: str, request: Request):
    """Delete a meal and update daily summary.

    Feature: 003-update-logic-for (Meal management)
    Cascades to photos and recalculates daily stats.
    """
    user_id = await get_authenticated_user_id(request)

    # Validate meal exists and ownership
    try:
        existing_meal = await db_get_meal_with_photos(UUID(meal_id))
    except ValueError:
        raise HTTPException(status_code=422, detail="Invalid meal ID format") from None

    if not existing_meal:
        raise HTTPException(status_code=404, detail="Meal not found")

    if str(existing_meal.userId) != user_id:
        raise HTTPException(status_code=403, detail="You do not own this meal")

    # Delete meal (cascades to photos, updates daily summary)
    success = await db_delete_meal(meal_id)  # db_delete_meal expects str
    if not success:
        raise HTTPException(status_code=500, detail="Failed to delete meal")

    return None  # 204 No Content

File: frontend/src/services/meals.ts
/**
 * Meals Service
 * Feature: 003-update-logic-for
 * Task: T054
 *
 * API client for meal management operations
 */

import { api } from './api';

export interface Macronutrients {
  protein: number;
  carbs: number;
  fats: number;
}

export interface Photo {
  id: string;
  thumbnailUrl: string;
  fullUrl: string;
  displayOrder: number;
}

export interface Meal {
  id: string;
  userId: string;
  createdAt: string;
  description: string | null;
  calories: number;
  macronutrients: Macronutrients;
  photos: Photo[];
  confidenceScore: number | null;
}

export interface MealUpdate {
  description?: string;
  protein_grams?: number;
  carbs_grams?: number;
  fats_grams?: number;
}

export interface MealsListResponse {
  meals: Meal[];
  total: number;
}

export interface CalendarDay {
  meal_date: string;
  meal_count: number;
  total_calories: number;
  total_protein: number;
  total_carbs: number;
  total_fats: number;
}

export interface MealsCalendarResponse {
  dates: CalendarDay[];
}

/**
 * Get meals for a specific date or date range
 * Feature: 003-update-logic-for
 */
export async function getMeals(date?: string, startDate?: string, endDate?: string, limit: number = 50): Promise<MealsListResponse> {
  const params = new URLSearchParams();

  if (date) {
    params.append('date', date);
  } else if (startDate && endDate) {
    params.append('start_date', startDate);
    params.append('end_date', endDate);
  }

  if (limit) {
    params.append('limit', limit.toString());
  }

  const url = `/api/v1/meals${params.toString() ? `?${params.toString()}` : ''}`;
  const response = await api.get<MealsListResponse>(url);
  return response.data;
}

/**
 * Get a specific meal by ID
 * Feature: 003-update-logic-for
 */
export async function getMeal(mealId: string): Promise<Meal> {
  const response = await api.get<Meal>(`/api/v1/meals/${mealId}`);
  return response.data;
}

/**
 * Update meal details
 * Feature: 003-update-logic-for
 */
export async function updateMeal(mealId: string, updates: MealUpdate): Promise<Meal> {
  const response = await api.patch<Meal>(`/api/v1/meals/${mealId}`, updates);
  return response.data;
}

/**
 * Delete a meal
 * Feature: 003-update-logic-for
 */
export async function deleteMeal(mealId: string): Promise<void> {
  await api.delete(`/api/v1/meals/${mealId}`);
}

/**
 * Get calendar summary for date range
 * Feature: 003-update-logic-for
 */
export async function getMealsCalendar(startDate: string, endDate: string): Promise<MealsCalendarResponse> {
  const params = new URLSearchParams({
    start_date: startDate,
    end_date: endDate,
  });

  const response = await api.get<MealsCalendarResponse>(`/api/v1/meals/calendar?${params.toString()}`);
  return response.data;
}

export default {
  getMeals,
  getMeal,
  updateMeal,
  deleteMeal,
  getMealsCalendar,
};
